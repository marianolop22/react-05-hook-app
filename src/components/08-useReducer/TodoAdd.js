import React from 'react'
import { useForm } from '../../hooks/useForm'

export const TodoAdd = ({handleAddTodo}) => {

    //desestructura el value, ya que es un objeto y toma solo el campo description
    const [ {description}, handleInputChange, reset ] = useForm ({
        description: ''
    })

    const handleSubmit = (e) => {
        
        e.preventDefault();

        if ( description.trim() <=1 ) {
            return;
        }
        const newTodo = {
            id: new Date().getTime(),
            desc: description,
            done: false
        };

        handleAddTodo(newTodo);
        reset();
    };


    return (
        <>
            <h4>Agregar TODO</h4>
            <hr/>
            <form onSubmit={handleSubmit}>
                <input
                    className="form-control"
                    text="text"
                    name="description"
                    placeholder="aprender...."
                    autoComplete="off"
                    value={description}
                    onChange={handleInputChange}
                />
                <button
                    type="submit"
                    className="btn btn-primary mt-1 btn-block"
                >
                    Agregar
                </button>
            </form>
        </>
    )
}
