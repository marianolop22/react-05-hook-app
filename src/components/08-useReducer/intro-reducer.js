
const initialState = [{
    id: 1,
    todo: 'Comprar pan',
    done: false
}];

//el reducer labura asÃ­. recibe 2 parametros, estado  y funcion
const todoReducer = ( state=initialState, action ) => {

    if ( action?.type === 'agregar') {
        return [...state, action.payload]; //esto es como un push

    }


    return state;
};

let todos = todoReducer();

const newTodo = {
    id: 2,
    todo: 'Comprar leche',
    done: false
};

//se crea una accion, como estandard debe tener la propiedad type
const agregarTodoAction = {
    type: 'agregar',
    payload: newTodo //payload es un estandard
};

todos = todoReducer ( todos, agregarTodoAction);





console.log(todos);





