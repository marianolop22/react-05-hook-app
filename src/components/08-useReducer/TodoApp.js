import React, { useEffect, useReducer } from 'react'

import './styles.css';
import { TodoAdd } from './TodoAdd';
import { TodoList } from './TodoList';
import { todoReducer } from './todoReducer';

const init = () => {
    return JSON.parse( localStorage.getItem('todos') ) ||  [];
    // [{
    //     id: new Date().getTime(),
    //     desc: 'Aprender React',
    //     done: false
    // }];
};

export const TodoApp = () => {

    //reducer: es la funcion, initial state: es el estado, init es la funcion que ayuda a computar el estado inicial
    const [todos, dispach] = useReducer(todoReducer, [], init);

    useEffect(() => {
        localStorage.setItem('todos', JSON.stringify(todos));
    }, [todos])

    const handleDelete = (todoId) => {
        const action = {
            type:'delete',
            payload: todoId
        };
        dispach(action);
    };

    const handleToggle = ( todoId ) => {
        const action = {
            type:'toggle',
            payload: todoId
        };
        dispach(action);
    };

    const handleAddTodo = (newTodo) => {
        const action = {
            type:'add',
            payload: newTodo
        };
        dispach(action);
    };



    return (
        <div>
            <h1>TodoApp ({todos.length})</h1>
            <hr/>

            <div className="row">
                <div className="col-7">
                    <TodoList 
                         todos={todos} 
                         handleDelete={handleDelete} 
                         handleToggle={handleToggle} 
                    />
                </div>
                <div className="col-5">
                    <TodoAdd handleAddTodo={handleAddTodo}/>

                </div>

            </div>
        </div>
    )
}
