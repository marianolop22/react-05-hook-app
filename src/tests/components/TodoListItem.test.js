import '@testing-library/jest-dom';
import React from 'react';
import { shallow } from 'enzyme';
import { TodoListItem } from "../../components/08-useReducer/TodoListItem";
import { demoTodos } from "../fixtures/demoTodos";


describe ('Pruebas de TodoListItem', ()=> {

    const handleDelete = jest.fn();
    const handleToggle = jest.fn();
    const index=1;
    const wrapper = shallow(
            <TodoListItem 
                todo={demoTodos[1]} 
                index={index} 
                handleDelete={handleDelete} 
                handleToggle={handleToggle}
            />);


    test('debe renderizar correctamente', () => {
        expect(wrapper).toMatchSnapshot();
    });

    test('debe de llamar la funcion borrar', () => {
        wrapper.find('button').simulate('click');
        expect(handleDelete).toHaveBeenCalledWith(demoTodos[1].id);
    });

    test('debe de llamar la funcion toggle', () => {
        wrapper.find('p').simulate('click');
        expect(handleToggle).toHaveBeenCalledWith(demoTodos[1].id);
    });

    test('debe de mostrar el texto correctamente', () => {
        expect(wrapper.find('p').text()).toBe(`${index+1}. ${demoTodos[1].desc}`);
    });

    test('debe de tener la clase complete si el todo.done es true', () => {
        
        const todo=demoTodos[0];
        todo.done=true;

        const wrapper = shallow(
            <TodoListItem 
                todo={todo} 
                index={0} 
                handleDelete={handleDelete} 
                handleToggle={handleToggle}
            />);

        expect (wrapper.find('p').hasClass('complete')).toBe(true);
    });
});