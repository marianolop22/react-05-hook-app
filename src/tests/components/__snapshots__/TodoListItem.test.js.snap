// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas de TodoListItem debe renderizar correctamente 1`] = `
<li
  className="list-group-item"
  key="2"
>
  <p
    className="false"
    onClick={[Function]}
  >
    2
    . 
    Otra cosa
  </p>
  <button
    className="btn btn-danger"
    onClick={[Function]}
  >
    Borrar
  </button>
</li>
`;
